docs: ADR-4: max collateral utilization (#940)
umee-network/umee: docs/architecture/ADR-004-interest-and-reserves.md

## Decision

Interest on all open borrow positions will be applied every X blocks during `EndBlock`, where X is a governance parameter called `BorrowInterestEpoch`.

Reserves are implemented as a per-token value attached to the `x/leverage` module account. The module account also holds the lending pool of base assets. As an example, if the module account contains 2000 `uatom` and its `uatom ReserveAmount` is 500, then only 1500 of its `uatom` are available for borrowing or withdrawal.

A governance parameter `ReserveFactor` must be kept which specifies the portion of borrow interest that will be funnelled into reserves. This parameter is set _per token_, which allows it to be set differently for riskier assets vs stablecoins, for example.

Both the reserved amount of a given token, and the token:uToken exchange rate, increase when interest is _accrued_, rather then when it is _repaid_.

This timing of reserve increases, matches the behavior of the [Compound cToken smart contract](https://github.com/compound-finance/compound-protocol/blob/master/contracts/CToken.sol) we are using as a reference - see around line 410 in function `accrueInterest`.

### Max collateral utilization

> We define a _collateral utilization_ as ratio of lended collateral to total amount of collateral per token. Collateral utilization is 0 when there is no borrow of given token collateral. It is equal to 1, when all token collateral is borrowed. Collateral utilization of token X is growing when lenders withdraw their token X collateral or borrowers take a new loan of token X.

Reviewer: Add some math or specifics here - is this _borrow utilization_ which is `borrowed / (borrowed + module balance - reserves)` or is it _collateral utilization_, which would have to distinguish between lenders with uTokens enabled vs. disabled as collateral?

Let's draw the follwing scenario to picture the liquidators risk:

1. Alice is providing \$1M USD for lending.
2. Bob is providing \$1.5M in Luna as a collateral and borrows 1M USD from Alice.
3. Charlie provides \$1.5M in BTC as a collateral and borrows $1M in Luna from Bob.
4. Charlie predicts Luna collapse and sells the Luna.
5. Luna is sinking and Bob position has to be liquidated. However:
   - Lenders can liquidate Bob, but they can only redeem up to 33% of `u/Luna` because the rest is not available (Charlie borrowed it).
   - Charlie will not pay off her borrow position - she will wait for the final collapse and buy Luna cheaply.
   - Liquidators will not take the risk of obtaining and holding `u/Luna` when there is a risk of Luna sinking deep.
6. In case of the big crash, liquidators won't perform a liquidation, Bob will run away with 1M USD, system will end up with a bad debt and obligation to pay Alice.

> We propose to set a per token parameter: **max collateral utilization** using Cosmos SDK v0.46 gov messages. The system will forbid to make any collateral related operation if the operation would move token _collateral utilization_ above _max collateral utilization_.

Reviewer: ```suggestion
We propose to add a per token parameter: **max collateral utilization** to the token registry. The system will forbid to make any collateral related operation if the operation would move token _collateral utilization_ above _max collateral utilization_.
```

}
```

> We modify the Compound formula to use collateral utilization rather than supply utilization, and interpolate over the collateral utilization to compute the borrow interest rate:

Reviewer: ```suggestion
```

}
```

In the codebase, the function above will be written more efficiently with regards to calculating utilization and interest rates to avoid the iteration over all user borrow positions per block. Error handling and iterator details are also omitted for clarity here.

> ### Protocol Reserve (insurance treasury)

Author: this is intentional - Reserve is not meaningful enough.

> Leverage module will build treasury to secure the protocol and act as a protocol insurance.

Reviewer: ```suggestion
The leverage module will build reserves to secure the protocol and act as insurance. These reserves are used during bad debt repayment, which is discussed in ADR 007.

```

Author: I wanted specifically to use word _treasury_ rather than _reserves_

Reviewer: Reserves are what they are called - in internal documentation, external API, code, etc.

If changing their name (everywhere), we should do it in a separate PR as this one has gotten off-topic.

### uToken Exchange Rate Impact

The token:uToken exchange rate, which would have previously been calculated by `(total tokens borrowed + module account balance) / uTokens in circulation` for a given token and associated uToken, must now be computed as `(total tokens borrowed + module account balance - reserved amount) / uTokens in circulation`.

> Also, because Lend, Withdraw, Borrow, and Repay transactions don't affect the token:uToken exchange rate, it's enough to update the exchange rate during the EndBlocker `AccrueAllInterest`.

Reviewer: ```suggestion
```
We can remove this line altogether since exchange rate is now computed on demand, instead of being stored.

Author: I think this is paragraph is meaningful.

The exchange rate is stored under the following key:

```
// append 0 for null-termination
exchangeRatePrefix | denom | 0

> ```

Reviewer: Needs to be removed since exchange rate no longer stored

Author: this was already there, I moved it to a common place.
We need to store somewhere the exchange rates.

Reviewer: Exchange rate is currently not stored in state - it is computed on demand when needed for queries/transactions.

Author: OK, so let's remove it.

> The global `ReserveFactor` paramater is 0.05
>
> The derived interest rate for `atom` is assumed to be 0.0001% (10^-6) per block.
>
> Alice has borrowed 2000 `atom`, which is 2 \* 10^9 `uatom` internally. (Note: These are not utokens, just the smallest units of the token. 1 atom = 10^6 uatom.)
>
> During the next `EndBlock`, interest is accrued on the borrow. Alice's borrowed position is increased by the interest: now she owns 2.000002 `uatom`. It increases by 2000 `uatom`.
> The `x/leverage` module's reserved amount of `uatom` increases by 100 ( `2000 * ReserveFactor = 2000 * 0.05 = 100` 2000 \* 0.05) due to the interest accruing on Alice's loan.
>
> The same math is applied to all open borrows, which may be from different users and in different asset types, during EndBlock.

> Note that it is the module "reserved amount" has increased, not the actual balance of the `x/leverage` account.

Reviewer: ```suggestion
Note that the module "reserved amount" has increased, but not the actual balance of the `x/leverage` account.
```
